WIP: Add support for rendering Markdown contents

khos2ow opened the pull request:
### Prerequisites

Put an `x` into the box(es) that apply:

- [ ] This pull request fixes a bug.
- [x] This pull request adds a feature.
- [ ] This pull request introduces breaking change.

For more information, see the [Contributing Guide](https://github.com/segmentio/terraform-docs/tree/master/CONTRIBUTING.md).

### Description

Generate Inputs (Required and Optional) and Outputs and show them in markdown _section_ format. (i.e. different heading). This feature shares a lot of its code with `markdown` code so I extracted some of the shared code to other files to be able to reuse them in both places.

### Issues Resolved

List any existing issues this pull request resolves.

### Checklist

Put an `x` into all boxes that apply:

- [x] I have read the [Contributing Guidelines](https://github.com/segmentio/terraform-docs/tree/master/CONTRIBUTING.md).

#### Tests

- [x] I have added tests to cover my changes.
- [x] All tests pass when I run `make test`.

#### Documentation

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

#### Code Style

- [x] My code follows the code style of this project.

khos2ow commented:
ping @metmajer for review.

metmajer commented:
Hi @khos2ow and thank you for your contribution!

There's a couple of things that need some further elaboration:

1) Since they're so closely related, I'd rather like to have a `terraform-docs md --table` and `terraform-docs md --document` instead of `terraform-docs md` and `terraform-docs section`. Obviously, we'd have `terraform-docs md` default to `terraform-docs md --table` for backwards compatibility.
This will require an update of the `docopts` library, which I will take care of in a separate story.
2) You've created functions `GetNormalizedDescription()` in `pkg/doc/input.go` and `pkg/doc/output.go`, but I am not a fan of these. This normalization deals with representation, and therefore shouldn't go into the `pkg/doc` package, which represents the data layer.
3) There are comments including `TODO`, but we'll sort these out as we advance with the review.

Here's a suggestion: your points with refactoring functionality into `pkg/print` are valid, but I'd like to handle refactorings a little differently. In order to facilitate your contribution, please undo any refactorings to the code and simply incorporate your feature into the current architecture, even if this entails code duplications. Cheers!

khos2ow commented:
@metmajer Thanks for the review.

- I reverted item 2, 3 and `pkg/print`
- I didn't use `--document` and used `--content` instead, mainly because there already is a concept of `doc`, so I create `content.go`
- I changed the structure of `makrdown` folder a bit to have two separate subfolders `table` and `content`

Please let me know if you have any comments or suggestions on any of the above.

khos2ow commented:
@metmajer any more changes required for this PR?

metmajer commented:
Hi @khos2ow. Apologies, I've been busy and traveling for the past weeks.

Yes, there are things we should address and I'll need some more time to have everything documented.

khos2ow commented:
Thanks @metmajer, I wouldn't expect otherwise, having a clean and consistent repo is the most important thing to consider. I'm wondering if you forgot to submit the review or you meant that you're going to send the comments and suggestions at a later time?

As for the only comment, regarding `internal/pkg/print/markdown/content/content.go` I totally see your point, my reason to moved those `func` to the bottom of the file was that those are more like _private_/_helper_ functions for all the other `printXXX` functions. I kept more important functions at top and the rest at the bottom, both in alphabetic order. But that for sure can change :)

metmajer commented:
Hey @khos2ow. You should find a trail of code highlights and comments for your changes right above my comment. This may not be visible from your mobile phone. Please verify that you can see these comments!

khos2ow commented:
No I cannot see anything. That's why I suspected you actually forgot to _submit_ your review.

metmajer commented on README.md:
```diff
@@ -28,7 +28,7 @@ This project is no longer maintained by Segment. Instead, [Martin Etmajer](https
 
   Usage:
-    terraform-docs [--no-required] [--no-sort | --sort-inputs-by-required] [--with-aggregate-type-defaults] [json | markdown | md] <path>...
+    terraform-docs [--no-required] [--no-sort | --sort-inputs-by-required] [--with-aggregate-type-defaults] [--markdown-table | --markdown-content] [json | markdown | md] <path>...
```
The position of these options should be fixed. Right now, it seems they are global properties that can be applied to any of `json | markdown | md`, whereas they should only be taken into account when the user has selected either `markdown` or `md`. Please see [github.com/docopt/docopt-go](github.com/docopt/docopt-go) for help.

metmajer replied:
The name and position of arguments would then also need to be adapted in other places in `README.md` as well as the documentation header in `main.go`.

khos2ow replied:
That was my first approach: `[ json | markdown | md ] [ table | content ]` with correct dependency of subcommand, but I wasn't able to make it work through the `docopt`. To be honest I don't remember what was the issue, but I'll give it another try to see.

metmajer replied:
This might be because we're using an outdated library. There's an issue #82 to upgrade to the official repository. I have this assigned to myself, but feel free to take ownership!

khos2ow replied:
There might be a problem with flags, right now I changed to `terraform-docs md table ...` (as opposed to `... md --markdown-table ...`) and it works.

metmajer replied:
Using subcommands is a great choice. Right now, those subcommands would also apply to `json`. Is there a better way of structuring commands / subcommands so that `table` and `document` only apply to `markdown` and `md`?

metmajer replied:
Right, but independent of the documentation, the usage suggests, that these subcommands are available with `json` as well. By the way, I've just pushed #91 to update to the latest revision of `github.com/docopt/docopt.go`.

khos2ow replied:
Yes, that's true. I'm not sure what I can do on that regard yet. I'll try to see what can be done.

khos2ow replied:
Considering [this](http://try.docopt.org/?doc=Usage%3A%0D%0A++terraform-docs+%5Boptions%5D+json+%3Cpath%3E...%0D%0A++terraform-docs+%5Boptions%5D+%5Bmd+%7C+markdown%5D+%5Btable+%7C+document%5D+%3Cpath%3E...%0D%0A++terraform-docs+-h+%7C+--help%0D%0A%0D%0AOptions%3A%0D%0A++-h%2C+--help+++++++++++++++++++++++show+help+information%0D%0A++--no-required++++++++++++++++++++omit+%22Required%22+column+when+generating+markdown%0D%0A++--no-sort++++++++++++++++++++++++omit+sorted+rendering+of+inputs+and+ouputs%0D%0A++--sort-inputs-by-required++++++++sort+inputs+by+name+and+prints+required+inputs+first%0D%0A++--with-aggregate-type-defaults+++print+default+values+of+aggregate+types%0D%0A++--version++++++++++++++++++++++++print+version%0D%0A%0D%0ATypes+of+markdown+%28optional%29%3A++++++table+will+be+used+if+not+specified%0D%0A++table++++++++++++++++++++++++++++generate+markdown+table+of+document%0D%0A++document+++++++++++++++++++++++++generate+markdown+document+of+document%0D%0A&argv=json+md+..%2Fhome), [this](http://try.docopt.org/?doc=Usage%3A%0D%0A++terraform-docs+%5Boptions%5D+json+%3Cpath%3E...%0D%0A++terraform-docs+%5Boptions%5D+%5Bmd+%7C+markdown%5D+%5Btable+%7C+document%5D+%3Cpath%3E...%0D%0A++terraform-docs+-h+%7C+--help%0D%0A%0D%0AOptions%3A%0D%0A++-h%2C+--help+++++++++++++++++++++++show+help+information%0D%0A++--no-required++++++++++++++++++++omit+%22Required%22+column+when+generating+markdown%0D%0A++--no-sort++++++++++++++++++++++++omit+sorted+rendering+of+inputs+and+ouputs%0D%0A++--sort-inputs-by-required++++++++sort+inputs+by+name+and+prints+required+inputs+first%0D%0A++--with-aggregate-type-defaults+++print+default+values+of+aggregate+types%0D%0A++--version++++++++++++++++++++++++print+version%0D%0A%0D%0ATypes+of+markdown+%28optional%29%3A++++++table+will+be+used+if+not+specified%0D%0A++table++++++++++++++++++++++++++++generate+markdown+table+of+document%0D%0A++document+++++++++++++++++++++++++generate+markdown+document+of+document%0D%0A&argv=md+json+..%2Fhome) and [this](http://try.docopt.org/?doc=Usage%3A%0D%0A++terraform-docs+%5Boptions%5D+json+%3Cpath%3E...%0D%0A++terraform-docs+%5Boptions%5D+%5Bmd+%7C+markdown%5D+%5Btable+%7C+document%5D+%3Cpath%3E...%0D%0A++terraform-docs+-h+%7C+--help%0D%0A%0D%0AOptions%3A%0D%0A++-h%2C+--help+++++++++++++++++++++++show+help+information%0D%0A++--no-required++++++++++++++++++++omit+%22Required%22+column+when+generating+markdown%0D%0A++--no-sort++++++++++++++++++++++++omit+sorted+rendering+of+inputs+and+ouputs%0D%0A++--sort-inputs-by-required++++++++sort+inputs+by+name+and+prints+required+inputs+first%0D%0A++--with-aggregate-type-defaults+++print+default+values+of+aggregate+types%0D%0A++--version++++++++++++++++++++++++print+version%0D%0A%0D%0ATypes+of+markdown+%28optional%29%3A++++++table+will+be+used+if+not+specified%0D%0A++table++++++++++++++++++++++++++++generate+markdown+table+of+document%0D%0A++document+++++++++++++++++++++++++generate+markdown+document+of+document%0D%0A&argv=md+table+..%2Fhome) I think we won't be able to only depends on docopt and we need to do some internal sanity check too. The same argument can be said for mutually exclusive `--no-sort` and `--sort-inputs-by-required`. So the Usage can be shorten to something like in the example links above.

And even considering the upcoming `v1.0.0` release even some of those options can be even shortened (`--aggregate`, etc.)

metmajer commented on internal/pkg/print/markdown/content/content.go:
```diff
@@ -0,0 +1,144 @@
+	return buffer.String(), nil
+}
+
+func printComment(buffer *bytes.Buffer, comment string, settings settings.Settings) {
```
Please retain the order alphabetical order of functions.

khos2ow replied:
+1

metmajer commented on internal/pkg/print/markdown/content/testdata/content-WithAggregateTypeDefaults.golden:
```diff
@@ -0,0 +1,116 @@
+
+Type: `string`
+
+Default: -
```
Same as above. Not sure if in a non-table layout, documenting non-existent default value is of much use. By the way, I suggest you create another variable named `string-3` with `default = ""` to have this use-case covered (which becomes a useful test scenario with this feature).

khos2ow replied:
There's one fundamental difference between `document` and the rest of the print types. In `document` when `WithRequired` is selected there will be two different sections:

- `Required Inputs`: doesn't show `Default: -` at all (there's no default value for required inputs)
- `Optional Inputs`: does show `Default: <value>`

But without `--no-required` there will only be one sections, `Inputs`. This is identical to `table` or `json` in terms of showing `Default: -`. I think to make it still consistent within `document` type we can both:

- hide `Default: -` for non-existing `default` Terraform attribute
- show ``Default: `""` `` for _empty_ `default` Terraform attribute

any thought?

khos2ow replied:
any thought on this?

metmajer replied:
I like what you suggest. As said before, we haven't captured the use-case of empty strings in `variables.tf` yet, which we should definitely do. A quick test shows... 

1) that a variable that defaults to an empty string results in `` in the table layout
2) compare this with `string-1`, which results in `bar`, not `"bar"`

I'd say that `` isn't as obvious a depiction of an empty string as `""` is. The JSON output of `--with-aggregate-type-defaults` can be regarded as an extension to the default view that only shows concrete default values for variables of type `string`, such as `bar`. If the variable was a `map`, we would show `{ "foo": "bar" }`, so why not show defaults of type strings as `"bar"` instead of `bar` in all markdown documents?

metmajer replied:
Suggestion: to keep this feature focused, I'd produce Default: `` now and later provide a fix that turns `` into `""` for all markdown layouts (please look at the source of this comment to understand what I mean).

metmajer commented on main.go:
```diff
@@ -83,10 +94,12 @@ func main() {
-		out, err = markdown.Print(document, printSettings)
+	case args["markdown"].(bool), args["md"].(bool):
+		if args["--markdown-content"].(bool) {
+			out, err = content.Print(document, printSettings)
```
Please use aliases `markdown_document` and  `markdown_table` in your package imports and re-use them here so that it reads `markdown_document.Print(...)`, etc..

khos2ow replied:
+1

khos2ow commented on AUTHORS:
```diff
@@ -1,13 +1,16 @@
 Sergiusz Urbaniak <sergiusz.urbaniak@gmail.com>
-Stuart Auld <sja@marsupialmusic.net>
+Simon Heather <metmajer@getcloudnative.io>
+Stuart Auld <metmajer@getcloudnative.io>
```
There actually is a problem in `authors` target which #90 fixes. I'll leave this alone for now and rebase later when the other PR gets merged.

metmajer replied:
Just merged #90.

khos2ow replied:
Rebased. This should be gone now.

khos2ow commented on README.md:
````diff
@@ -127,6 +136,33 @@ This module has a variable and an output.  This text here will be output before
+```bash
+$ terraform-docs md --markdown-content _example
+This module has a variable and an output.  This text here will be output before any inputs or outputs!
+
````
I agree but:

- the extra empty line is inside a markdown fence, so it doesn't matter for linting
- this line was added in ccf85a95df9c55b42ba2e2f6f3e8af8b254f2b2a and it has been like this for quite a while (and I just duplicated the `table` example`
- and actually it's inconsistent with the actual `./_example/main.tf` content

I would suggest we fix this in a separate PR, because potentially we need to fix all the examples in `README` based on 1) content of `./_example/main.tf` or 2) provided [Example](https://github.com/segmentio/terraform-docs/blob/master/README.md#example) in `README`

metmajer replied:
Agreed. I wasn't aware that this a) was already in there and b) the linter doesn't care. Thanks!

khos2ow commented on main.go:
```diff
@@ -17,7 +18,7 @@ var version = "dev"
 const usage = `
   Usage:
-    terraform-docs [--no-required] [--no-sort | --sort-inputs-by-required] [--with-aggregate-type-defaults] [json | markdown | md] <path>...
+    terraform-docs [--no-required] [--no-sort | --sort-inputs-by-required] [--with-aggregate-type-defaults] [--markdown-table | --markdown-content] [json | markdown | md] <path>...
```
+1

khos2ow commented on main.go:
```diff
@@ -7,7 +7,8 @@ import (
 	"github.com/segmentio/terraform-docs/internal/pkg/print"
 	"github.com/segmentio/terraform-docs/internal/pkg/print/json"
-	"github.com/segmentio/terraform-docs/internal/pkg/print/markdown"
+	markdown_content "github.com/segmentio/terraform-docs/internal/pkg/print/markdown/content"
```
btw, I'm going to rename and `git mv` the `markdown/conent` to `markdown/document` at the very last for the purpose for code review be easier.

metmajer replied:
Agreed, please continue with the renaming.

khos2ow replied:
Renamed and moved to correct package name.

khos2ow commented on internal/pkg/print/markdown/content/testdata/content-WithAggregateTypeDefaults.golden:
```diff
@@ -0,0 +1,116 @@
+
+### unquoted
+
+Description: -
```
I personally like to still show `Description: -` (or `Description: not available` or similar), to make the blocks consistent. Otherwise some block would have description some wouldn't. And also we cannot remove empty description from  `table` or `json` format, so I think we should keep it for `document` type as well.

khos2ow replied:
any thought on this?

metmajer replied:
I see a couple of options here:
1) "Description: This is about..." and "Description: -"
2) "Description: This is about..." and "Description: n/a"
3) "This is about..." and "" (nothing)
Which one is your favorite? I'll leave this up to you.

khos2ow replied:
I personally like option 2, both for markdown `document` and `table` (specially for consistency). I also believe this should remain as is in `json` because of potential integration might have been setup on it already and also having an _empty_ item in a json object is perfectly normal.

metmajer replied:
Agreed. For the sake of keeping this feature branch lean, why not go with option 1 for consistency and then provide another update that provides 2) across all markdown implementations?

khos2ow replied:
Makes sense, I will do option 1 here and open a PR for option 2 later.

khos2ow replied:
I'm gonna resolve this since the feature currently works as is.

khos2ow commented on internal/pkg/print/markdown/content/testdata/content-WithAggregateTypeDefaults.golden:
```diff
@@ -0,0 +1,116 @@
+
+Type: `map`
+
+Default: `{ "a": 1, "b": 2, "c": 3 }`
```
I had considered this but the only reason I didn't do this was that it could have looked weird (the fact that the fence is _under_ "Default" not _in front of_ it:

---

### variable-1

Description: value-1 description

Type: `string`

Default: `foo`

### variable-2

Description: value-2 description

Type: `map`

Default: 
```
{
  "a": 1,
  "b": 2,
  "c": 3
}
```

### variable-3

Description: value-3 description

Type: `string`

Default: `bar`

---

Or may be I'm reading too much into this! :wink: It doesn't actually look very bad!

metmajer replied:
I like it pretty much, actually :)

khos2ow replied:
Sure, pretty-printed code fence it is.

khos2ow replied:
This should be prettified now.

khos2ow commented:
@metmajer do you think anything is still missing for this PR?

metmajer commented:
Thank you for your patience, @khos2ow. I'll get back to you with an update in the next days.

metmajer commented:
Hi @khos2ow. I had a list of change requests which would have been hard to communicate. Therefore, I code them up myself and filed a pull request to your branch at github.com/khos2ow/terraform-docs:content. If you could accept, we'd be good to go, but please go have a look first. Thanks!

khos2ow commented:
Cool, but I could have enabled `Allow edits from maintainers` on this PR for you. Anyway I put a small comment on your PR.

khos2ow commented:
Thank you @metmajer for your support and review! :tada:
